test/node/node_test.cpp

```diff
@@ -91,6 +97,30 @@ TEST(NodeTest, MapIteratorWithUndefinedValues) {
   EXPECT_EQ(1, count);
 }
 
+TEST(NodeTest, ConstIteratorOnConstUndefinedNode) {
+  Node node;
+  const Node& cn = node;
+  const Node& undefinedCn = cn["undefined"];
+
+  std::size_t count = 0;
+  for (const_iterator it = undefinedCn.begin(); it != undefinedCn.end(); ++it)
+    count++;
+  EXPECT_EQ(0, count);
+}
+
+TEST(NodeTest, IteratorOnConstUndefinedNode) {
+  Node node;
+  const Node& cn = node;
+  const Node& undefinedCn = cn["undefined"];
+
+  Node& nonConstUndefinedNode = const_cast<Node&>(undefinedCn);
```

Comment:
I believe so - as it's only "const Node Node::operator[](const Key& key) const" that returns a 'zombie' node - "Node Node::operator[](const Key& key)", however, will always create a new node, with a NULL node* - which does not hit the if(!m_isValid) exception.

Then , as far as I know, the const_cast<> is required to get a non-const iterator out of the undefined 'zombie' node returned from the const operator[] version. (Though I see it as if something breaks because of something you can only get access to by const_cast<>ing a value to non-const it's kinda your own fault - I only did the same for the non-const interator functions for completeness)